Implement date-field element

adamstep opened the pull request:
[TAD](https://instawork.atlassian.net/wiki/spaces/EN/pages/901677061/Date+field+element)

#### Feature
Implements the `<date-field>` element for selecting dates. This works similar to the `<picker>` field:
- renders a field (view with text) on the screen. The view and text can be styled.
- tapping the field focuses it and brings up the date selector:
    - on iOS, we show a bottom sheet with the spinning date picker
    - on Android, we show the system date picking modal
- Selected dates are shown in the field, rendered in the given format.

#### Implementation details
- The date formatting (for the field) is handled via dependency injection from the client. In the demo app, we use the `moment` library to demonstrate the formatting.
- The date formatting function is passed to date fields from the root screen via a React context.

#### To-do
- [x] Android implementation
- [ ] Update website docs

#### Screenshots
iOS:
![date-field](https://user-images.githubusercontent.com/1034502/60410789-c1232000-9b7e-11e9-99fa-79a2105df584.gif)

divyanshu013 commented on src/components/hv-date-field/index.js:
```diff
@@ -0,0 +1,321 @@
+  };
+
+  toggleSavePress = () => {
+    this.setState({ donePressed: !this.state.donePressed });
```
Here as well:

```suggestion
    this.setState(prevState => ({ donePressed: !prevState.donePressed }));
```

divyanshu013 commented on src/components/hv-date-field/index.js:
```diff
@@ -0,0 +1,321 @@
+    const element: Element = this.props.element;
+    this.setState({
+      focused: false,
+      value: this.state.pickerValue,
```
Minor: since this is a derived state as well, it should use a callback pattern to update state as well.

divyanshu013 commented on src/index.js:
```diff
@@ -126,6 +126,10 @@ export function text(element, stylesheets, onUpdate, options) {
 
+// Provides the date format function to use in date fields
+// in the screen.
+export const FormatDateContext = React.createContext();
```
Very minor: I think `DateFormatContext` is a better name since it provides the date format. FormatDate sounds like an action or function which formats the date.

adamstep commented:
@divyanshu013 thanks, I implemented all of your suggestions and added support for Android:
![android](https://user-images.githubusercontent.com/1034502/60462718-eb1c2700-9bfe-11e9-846f-24307f11bf80.png)

adamstep commented on src/components/hv-date-field/index.js:
```diff
@@ -0,0 +1,355 @@
+    };
+  }
+
+  componentDidUpdate = async (prevProps: Props, prevState: State) => {
```
the API for the date picker on Android is an async function. RN docs recommend using it with the `await` syntax.
